#127 Adding basic procs that automatically fetches all "Many" entries of a "One-to-many" relationship

PhilippMDoerner opened the pull request:
This code adds basic procs to sqlite to fetch the other side of a "one-to-many" or a "many-to-many" relationship. It does so through generic procs that write the basic sql condition of the query for you on the related table you're fetching stuff on. In the case of many-to-many, it even unpacks the seq[joinTableModel] for you into seq[YourDesiredModel]. This unpacking is done via macro.

All the while, it manages to block you *at compile time* from using these procs incorrectly. If you try to use them with 2 models that do not have a many-to-one relationship, or 3 models that do not repesent a many-to-many relationship, your code will not compile.
`getRelatedFieldNameOn` figures that one out at compile time for you, as it acts almost entirely on the types.

The other 2 procs, `selectOneToMany` and `selectManyToMany` also figure out the sqlCondition to write at compile time, though the way they do it differs slightly between postgres and sqlite.

Particularly regarding naming of variables and more I'm very open to feedback, as at some point the ideas for naming things better just disappeared, which imo is very visible in my proc names but also variable names used within them. So I'm very much aware that this likely still needs some naming works, or where which could should be moved, but I believe this does encompass a basis to start from.

As for the direct feedback from the feature discussion:

> 1. Don't rely on newModel or similar “contructor” proc being defined. Instead, rely on object instances. There is no universal way to initialize a type in Nim, and even demanding a newModel proc to exist won't do: there are other ways to initialize an object. One is just call the type as proc, the other is how they do in in Status: proc init[T](_: typedesc[T]): T.

Done (That was more of a leftover from when I wrote this code in my own side-project and not quite intended for the norm library, I apologize for not cleaning that up)

> 2. It seems like the mapModel macro can be replaces with a template or even a plain proc?

That macro has been renamed `unpackFromJoinModel`. I tried a version of it as template before, but got stuck with that as I wanted to pass the field name as a string since that allowed me to use it with the output from `getRelatedFieldNameOn`. I'm not quite sure how else to do it, though would agree that a proc or template would be preferable, simply because they're easier to maintain.

> 3. Please use two spaces to indentation.

Done

> 4. let source = sourceType() won't work if sourceType has a DateTime in it. Again, use instances instead of types. I've already tried working with types in Norm 1, trust me, you don't want that :-)

For `getRelatedFieldNameOn` I sadly found no other way. Luckily, for `getRelatedFieldNameOn` that is actually perfectly fine, as it does not deal with field values and acts at compile time. I explicitly added `DateTime` to the `Doctor` model which I use in one of the tests, which passes. I did however update the two other procs (`selectOneToMany` and `selectManyToMany`) to work with model instances instead of types though.

> 5. You can use norm.model.isModel instead of cascading when to check is something is a Model.

I couldn't make a version of `getRelatedFieldNameOn` act at compile time while using `isModel`. I may have done something wrong, I can only say that I didn't manage to get that approach working while doing the things I wanted to do at compile time.

> 6. I'm not really seeing how mixin is useful here. Even Nim's manual states that mixin makes sense in templates and macros only ¯_(ツ)_/¯

Same as 1, that was on me and I fixed it accordingly. Sorry for the confusion I caused there.

PhilippMDoerner commented:
I've come the realization that there's an important limitation I overlooked:
`selectManyToMany` only works if your `joinModel` has the field you want to fetch with the actual model on them. It does *not* work if it's just an int64 field with the fk pragma. That is because `selectManyToMany` relies on norms already built in way of fetching related model instances for you. Norm doesn't do that if your field uses the fk pragma.

It might be a good idea to check for that and throw a sensible exception at compile time. It already errors out at compile time, thankfully, the error message is just comparatively cryptic. Here an example:

>  type mismatch: got 'seq[int64]' for 'manyEntries' but expected 'seq[Group]'

moigagoo commented:
@PhilippMDoerner Sorry for the slow response. I've been too busy to invest time in Norm lately.

Finally, I've found the time to review your PR. I've left some comments, please see. Note that I'm not repeating comments for SQLite if it already has been said for Postgres. Also, if I suggest adding dostrings to one proc, it means all new procs should have dostrings.

PhilippMDoerner commented:
Another unfortunate edge case is that this does not deal properly with models that have 2 foreign key fields to the same model.

`getRelatedFieldNameOn` will only give you the first field in the model with the foreign key. So if you wanted to get the related field entry from the second field, you would not be able to succeed.

This means that `getManyFromOne` must offer an additional (maybe optional) parameter for you to specify the field name with the desired foreign key pragma. 
It could still try to figure out the foreign key fieldname automatically for you if you don't provide a field name and throw an exception at compile time if there's more than one fieldname.

Alternatively there could be an overload for `getManyFromOne`, one where you can specify the foreign key field and the second one as the current one.

moigagoo commented:
@PhilippMDoerner 

>  I am not seeing any comments

I always forget to click Sumbit review 😞 Sorry.

PhilippMDoerner commented on src/norm/model.nim:
```diff
@@ -114,3 +114,25 @@ proc checkRo*(T: typedesc[Model]) =
     {.error: "can't use mutating procs with read-only models".}
 
+proc getRelatedFieldNameOn*[T: Model, M: Model](targetModel: typedesc[T], sourceModel: typedesc[M]): string  {.raises: [FieldDefect].} =
+  for sourceFieldName, sourceFieldValue in sourceModel()[].fieldPairs:
```
This contains another issue in that it won't deliver you what you'd expect if M has 2 foreign key fields to T. This doesn't let you properly pick the field if there are 2 fields and with the logic I provided that isn't exactly possible.
I stumbled over this issue a couple weeks ago during implementation in my nimstoryfont project where I already work with these procs.

The way around that is overloading "getManyFromOne" and "getManyToMany" so that you can pass them the name of the field you want to fetch stuff for. This will be included in a fairly large update to this PR where I'll also take your comments into account.

PhilippMDoerner replied:
1. Done
2. I went with "target" and "source", for I generally try to avoid using shorthands
3. Done.
4. I renamed it to T,S for target and source accordingly.

PhilippMDoerner commented on src/norm/postgres.nim:
```diff
@@ -316,3 +316,28 @@ template transaction*(dbConn; body: untyped): untyped =
+  const manyTableName: string = M.table()
+  const sqlCondition: string = manyTableName & '.' & foreignKeyFieldName & " = $1"
+
+  dbConn.select(relatedEntries, sqlCondition, oneEntry.id)
```
The intent was mostly for clarity. `M.table()` did not feel to me like it told you sufficiently that the tablename you're getting is of the "many" side from the one-to-many relationship. If you want it gone I can throw it out, not a big deal, just thought this way it would be clearer what you're doing and what you're getting, which means less thinking as you read over the code. and the goal of the `sqlCondition` becomes easier to understand.

PhilippMDoerner commented on src/norm/postgres.nim:
```diff
@@ -316,3 +316,28 @@ template transaction*(dbConn; body: untyped): untyped =
 
+
+# One-to-Many Fetching
+proc selectOneToMany*[O: Model, M: Model](dbConn; oneEntry: O, relatedEntries: var seq[M]) =
```
I fully agree on the docstring stuff, I totally slept on adding those as I was happy to just finally having found a solution, I'll add those post haste to all the included procs.

PhilippMDoerner replied:
Done

PhilippMDoerner commented on src/norm/model.nim:
```diff
@@ -114,3 +114,25 @@ proc checkRo*(T: typedesc[Model]) =
+
+  raise newException(
+    FieldDefect, 
+    "Tried getting foreign key field from model '" & name(sourceModel.type()) & "' to model '" & name(targetModel.type()) & "' but there is no such field!"
```
> Since strutils is already imported, you can replace that with a more readable "$# $#" % [foo, bar] syntax.

Funnily enough, since then I've also gone over to using formatstrings. Do you have a heavy preference for the $# syntax or would using the strformat lib also work for you (should have no performance impact since this all works at compile time)?

PhilippMDoerner replied:
@moigagoo 
I fixed that and restructured the proc a bit. Previously this would have very poorly understandable behaviour if you had a model A with 2 FK fields to model B. It would secretly hand you the first and ignore the second one. Now there's a way to specify which field name you want to use for the selection. `getRelatedFieldNameTo` is now only used for an optional convenience proc (an overload of `selectOneToMany`) if you don't want to specify that field name and want norm to infer it for you.

I used strformat for this one though as I personally prefer it over strutils formatting syntax.

PhilippMDoerner commented on src/norm/model.nim:
```diff
@@ -114,3 +114,25 @@ proc checkRo*(T: typedesc[Model]) =
   when T.hasCustomPragma(ro):
     {.error: "can't use mutating procs with read-only models".}
 
+proc getRelatedFieldNameOn*[T: Model, M: Model](targetModel: typedesc[T], sourceModel: typedesc[M]): string  {.raises: [FieldDefect].} =
```
> It takes two types, T and M. We assume M is an object type with a field of type T. The function returns the name of the field in M that has the type T.

Almost, it's capabilities are in fact a bit broader than that. The call `getRelatedFieldNameTo(S, T)`  returns the name of the field in S that "points to the same table that model T has". Which means this also works if S has a field with an fk-pragma that points to T. And it **also** works if S has a field of typeT2, if T2 has **the same tablename** as T via the `tableName` pragma. 

1. Using objects will not be possible here as the proc acts at compile time and I'm not aware how you would have objects at compile time for analysis. I'm personally insistent on keeping all of these calculations at compileTime, because the earlier you see these issues the better. If I can avoid a user running into an issue at runtime by having their code not compile when they do something wrong, I'd rather do that with a nice error message than let them notice at runtime. 
Regarding Datetime: Fascinatingly, it doesn't pose a problem. I assumed the same but tried it out in case it worked and found that DateTime does not cause issues here! Likely because this all acts at compileTime. I'll attach a code example to the bottom of this comment so you can run it and see it yourself. Here a [link to that code in play nim lang](https://play.nim-lang.org/#ix=3Q9n). 
2. You get the first field that matches, exactly as you assumed. As such I changed things around a bit. The core proc `selectOneToMany` now takes an additional parameter `foreignKeyFieldName` to circumvent *exactly* the problem you lined out there. This basically makes this proc almost obsolete. However, I added an overload of `selectOneToMany` that is without the parameter `foreignKeyFieldName` parameter and instead uses `getRelatedFieldNameTo` to infer it. You can still run into this scenario of 2 foreign key fields to the same model, however you'll get a FieldDefect at compile time with a nice error message if you do. See the test `"When there is multiple many-to-one relationships between two models and the type field for fetching the desired relationship is not specified, then do not compile"` (postgres/trelated.nim line 88).
3. Sadly not possible, doing so will cause the code to not compile, as the proc does have side-effects. I think the compiler counts the FieldDefects that might be raised as such side-effects.
4. I agree that the naming isn't great. `fieldsOfModelType` would be inaccurate though, as you only get a single foreign-key-field on the source model pointing to the target model. Strictly speaking even my name is accurate, since the model-field on `source` and the model specified in `target` don't even have to match, they merely need to point to the same table. Of course I could change that, but that feels like narrowing the capabilities of this proc down unnecessarily. Anyway, I agree that the name is bad, I am however also not satisfied with `fieldsOfModeltype`. If I name it how I use it I guess it would be `inferRelatedFieldNameToModel` or something like it, but I'm not convinced of that name either.


```nim
import norm/[model, pragmas]
import std/[typetraits, strformat, options, times, macros]


proc getRelatedFieldNameTo*[M: Model](targetTableName: static string, normModel: typedesc[M]): string {.compileTime.} =
  ## A compile time proc that searches the given `normModel` type for any 
  ## foreign key field that points to a table with the given name `targetTableName`. 
  ## Raises a FieldDefect at compile time if the model does not have exactly 
  ## one foreign key field to that table.
  var fieldNames: seq[string] = @[]
  const name = typetraits.name #Allows this generic proc to always have typetraits.name proc available even when the context it is called from doesn't import typetraits
  
  for sourceFieldName, sourceFieldValue in M()[].fieldPairs:
      #Handles case where field is an int64 with fk pragma
      when sourceFieldValue.hasCustomPragma(fk):
        when targetTableName == sourceFieldValue.getCustomPragmaVal(fk).table():
          fieldNames.add(sourceFieldName)
      
      #Handles case where field is a Model type
      elif sourceFieldValue is Model:
        when targetTableName == sourceFieldValue.type().table():
          fieldNames.add(sourceFieldName)
      
      #Handles case where field is a Option[Model] type
      elif sourceFieldValue is Option:
        when sourceFieldValue.get() is Model:
          when targetTableName == genericParams(sourceFieldValue.type()).get(0).table():
              fieldNames.add(sourceFieldName)

  if fieldNames.len() == 1:
    return fieldNames[0]
  
  elif fieldnames.len() < 1:
    let errorMsg = fmt "Tried getting foreign key field from model '{name(M)}' to model '{targetTableName}' but there is no such field!"
    raise newException(FieldDefect, errorMsg)
  
  elif fieldnames.len() > 1:
    let errorMsg = fmt "Can't infer foreign key field from model '{name(M)}' to model '{targetTableName}'! There is more than one foreign key field to that table!"
    raise newException(FieldDefect, errorMsg)

proc getRelatedFieldNameTo*[S: Model, T:Model](source: typedesc[S], target: typedesc[T]): string {.compileTime.} =
  ## A compile time proc that analyzes the given `source` model for any foreign key field that points to
  ## the table of the `target` model. Raises a FieldDefect at compile time if the model does not
  ## have exactly one foreign key field to that table 
  result = getRelatedFieldNameTo(T.table(), source)

type A* = ref object of Model
  time*: DateTime

type B* = ref object of Model
  somename*: string
  time*: DateTime
  fkToA*: A

type C* = ref object of Model
  da*: string
  time*: DateTime
  fkToA* {.fk: A.}: int64

type D* = ref object of Model
  du*: string
  time*: DateTime
  fkToA*: Option[A]


const val1 = B.getRelatedFieldNameTo(A)
const val2 = C.getRelatedFieldNameTo(A)
const val3 = D.getRelatedFieldNameTo(A)
echo val1
echo val2
echo val3
```

PhilippMDoerner commented on src/norm/postgres.nim:
```diff
@@ -316,3 +316,28 @@ template transaction*(dbConn; body: untyped): untyped =
     raise
 
+
+# One-to-Many Fetching
```
Done

PhilippMDoerner commented on src/norm/postgres.nim:
```diff
@@ -316,3 +316,28 @@ template transaction*(dbConn; body: untyped): untyped =
+
+# One-to-Many Fetching
+proc selectOneToMany*[O: Model, M: Model](dbConn; oneEntry: O, relatedEntries: var seq[M]) =
+  const foreignKeyFieldName: string = O.getRelatedFieldNameOn(M)
```
I don't think it would work at compile time in such a case, which is when `getRelatedFieldNameOn` acts (now renamed to `getRelatedFieldNameTo` and likely renamed again once we can agree on a name) .

PhilippMDoerner commented on src/norm/postgres.nim:
```diff
@@ -316,3 +316,28 @@ template transaction*(dbConn; body: untyped): untyped =
+
+# One-to-Many Fetching
+proc selectOneToMany*[O: Model, M: Model](dbConn; oneEntry: O, relatedEntries: var seq[M]) =
+  const foreignKeyFieldName: string = O.getRelatedFieldNameOn(M)
```
1. Does this still apply after the revision? 
2. I do this mostly in my own codebases since intellisense in vscode takes forever and I want to be able to see the types at glance if it's not immediately painfully obvious what they are. I removed them from "manyTableNames" and when defining "sqlCondition" since one could argue it *should* be obvious what type they return.
3. I covered this case by changing this proc to also receive a `foreignKeyFieldName` parameter and an overload that infer that parameter for me in a nieche scenario. One could argue if that nieche scenario is even desired or not. If not, I could throw out that overload as well as `getRelatedFieldNameTo`.

PhilippMDoerner commented:
@moigagoo Finally got around to writing some nim again. I went through the comments and adjusted what I could and fixed yours as well as my own suggestions. That is with the exception of my[ second comment](https://github.com/moigagoo/norm/pull/130#issuecomment-1013680427). Still haven't found something I'm 100% satisfied with to solve that one, though it's a minor gripe since at least it doesn't allow you to compile wrong code.

PhilippMDoerner commented on src/norm/postgres.nim:
```diff
@@ -316,3 +316,46 @@ template transaction*(dbConn; body: untyped): untyped =
+  ## a new seq of whatever type the field has.
+  newCall(bindSym"mapIt", mySeq, nnkDotExpr.newTree(ident"it", ident field))
+
+proc selectManyToMany*[M1: Model, J: Model, M2: Model](dbConn; queryStartEntry: M1, joinModelEntries: var seq[J], queryEndEntries: var seq[M2]) =    
```
Fixed this by adding 2 parameters that allow specifying the fields to use on the joinModel and provided a helper proc that keeps the API as before nice and simple

PhilippMDoerner commented:
With the merges of develop into this branch I would consider this done so far.
I had to do a couple partial imports because I need several procs from the macros lib that the new "pragmasutils" can not replace.

@moigagoo I'd request feedback or rejecting the PR if the code provided here is not moving in the direction you would want.
Should we agree on the code I'd move on to writing the docs for nim town.

moigagoo commented:
@PhilippMDoerner I'll be honest with you. I probably won't have time to review the PR any time soon. I'm from Russia, and things aren't going well. My entire life is about to change dramatically.

Really sorry to be the showstopper but I just can't invest any time in pet projects now.

PhilippMDoerner commented:
> @PhilippMDoerner I'll be honest with you. I probably won't have time to review the PR any time soon. I'm from Russia, and things aren't going well. My entire life is about to change dramatically.
> 
> Really sorry to be the showstopper but I just can't invest any time in pet projects now.

No worries, I can't claim to understand, I'm merely a horrified bystander for the most part of it. Wishing you the best and that hopefully soon peace will follow.

moigagoo commented:
@PhilippMDoerner could you please update the changelog with your changes? I'll merge the PR after that.

PhilippMDoerner commented:
@moigagoo Holy damn! Really nice to hear from you! Given the situation and your last comment I was starting to fear the worst!

Thank you very much for the response and especially for noticing the similarity in macro. Given the documentation on "dot", I think their intent is identical, but I believe they have subtle differences that become apparent during usage. I say that mostly because I tried replacing one with the other, but the tests didn't run through in that scenario. 

I'll take a look whether I can find a way to move this all to "dot" and spare ourselves the second macro (aka I'll ask on the discord because for the most part I still don't get macros. There's a reason I use compile time pragma on procs). If I can't make that possible, I'll likely move the "getField"-macro and "hasField" template into the "dot" file as they both perform very similar functions, seems better structurally.

Either way, I'll get to writing those changelog edits.

PhilippMDoerner commented:
@moigagoo I think the fact that dot is more specifically geared towards ref objects somehow makes it impossible to use "dot" instead of "getField".
Ironically though, if I just swap out "obj: ref object" with "obj: typed", I can get rid of "getField" easily enough.

PhilippMDoerner commented on src/norm/model.nim:
```diff
@@ -118,3 +118,82 @@ proc checkRo*(T: typedesc[Model]) =
+  return fieldNames[0]
+  
+
+macro getField(t: typed, fieldName: static string): untyped =
```
Good eye! I had to adjust the macro slightly (changed "ref object" to "typed"), but with that change I could use it and have all the tests run through flawlessly!

PhilippMDoerner commented:
Added changelog updates, should be all there then. Once this is merged I'll start another issue (and subsequently a PR) for writing docs
